vault/logical_system.go

```diff
@@ -2926,6 +2928,107 @@ func (b *SystemBackend) pathInternalUIResultantACL(ctx context.Context, req *log
 	return resp, nil
 }
 
+func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+
+	// Limit output to authorized paths
+	resp, err := b.pathInternalUIMountsRead(ctx, req, d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up target document and convert to map[string]interface{} which is what will
+	// be received from plugin backends.
+	doc := openapi.NewDocument()
+
+	procMountGroup := func(group, mountPrefix string) error {
+		for mount := range resp.Data[group].(map[string]interface{}) {
+			backend := b.Core.router.MatchingBackend(ctx, mountPrefix+mount)
+
+			if backend == nil {
+				continue
+			}
+
+			req := &logical.Request{
+				Operation: logical.HelpOperation,
+			}
+
+			resp, err := backend.HandleRequest(ctx, req)
+			if err != nil {
+				return err
+			}
+
+			var backendDoc *openapi.Document
+
+			// Normalize response type, which will be different if received
+			// from and external plugin.
+			switch v := resp.Data["openapi"].(type) {
+			case *openapi.Document:
+				backendDoc = v
+			case map[string]interface{}:
+				backendDoc = new(openapi.Document)
+				if err := mapstructure.Decode(v, backendDoc); err != nil {
+					return err
+				}
+			default:
+				continue
+			}
+
+			// Prepare to add tags to default builtins that are
+			// type "unknown" and won't already be tagged.
+			var tag string
+			switch mountPrefix + mount {
+			case "cubbyhole/", "secret/":
+				tag = "secrets"
+			case "sys/":
+				tag = "system"
+			case "auth/token/":
+				tag = "auth"
+			}
```

Comment:
Tags have no predefined meaning and can be applied per operation. They do provide things like openapi viewers some more info to group items, so for example our giant API is at least in "auth"/"system"/"secrets" groups when viewed. And logical mounts will get tagged "secrets", and credential mounts will get "auth". These default builtins are type "Unknown" so I've added special handling.

Reply:
This is a pure refactor that can probably wait, but I really don't like any tag handling happen at this level. Unfortunately, the straightforward way to tags some of these unknown backends is with mount location, which isn't part of the backend object. I considered passing it as extra data to the help request, but it would also bleed into changing `handleRootHelp` which felt pretty invasive. Any ideas on: 1. passing mount location so OpenAPI can use in 2. determining backend type another way?

If these backends are no longer of type "Unknown" (not sure why they're that way), all of this special handling can probably get ripped out.

Reply:
I think the mount location could be retrieved from the output of the ui/mounts response. But also i think this is fine as is, at least for now

Reply:
Can you also differentiate "identity"? It'll be moved out of secrets at some point (before 1.0 probably) as it really isn't a secrets engine.

Reply:
Yep, I can add an override for that. Do we expect in the future a new type will be added to:
```
const (
	TypeUnknown    BackendType = 0 // This is also the zero-value for BackendType
	TypeLogical    BackendType = 1
	TypeCredential BackendType = 2
)
```